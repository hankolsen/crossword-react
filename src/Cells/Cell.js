import React from 'react';
import PropTypes from 'prop-types';

import CellRectangle from './CellRectangle/CellRectangle';
import CellText from './CellText/CellText';
import ArrowAcross from './ArrowAcross';
import ArrowDown from './ArrowDown';
import CellLabel from './CellLabel/CellLabel';

const Cell = ({
  row,
  number = '',
  column,
  letter = '',
  arrow = '',
  clickHandler,
  highlighted,
  selected,
}) => (
  <g onClick={() => clickHandler(row, column)}>
    <CellRectangle
      column={column}
      row={row}
      selected={selected}
      highlighted={highlighted}
    />
    {number ? <CellLabel column={column} row={row} number={number} /> : null}
    {arrow && arrow === 'across' ? (
      <ArrowAcross column={column} row={row} />
    ) : null}
    {arrow && arrow === 'down' ? <ArrowDown column={column} row={row} /> : null}
    <CellText row={row} column={column} text={letter} />
  </g>
);

Cell.defaultProps = {
  number: '',
  letter: '',
  arrow: '',
};

Cell.defaultProps = {
  highlighted: false,
  selected: false,
};

Cell.propTypes = {
  row: PropTypes.number.isRequired,
  column: PropTypes.number.isRequired,
  number: PropTypes.number,
  letter: PropTypes.string,
  arrow: PropTypes.string,
  clickHandler: PropTypes.func.isRequired,
  highlighted: PropTypes.bool,
  selected: PropTypes.bool,
};

export default Cell;
