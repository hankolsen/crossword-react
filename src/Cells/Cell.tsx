import React from 'react';

import CellRectangle from './CellRectangle/CellRectangle';
import CellText from './CellText/CellText';
import ArrowAcross from './ArrowAcross';
import ArrowDown from './ArrowDown';
import CellLabel from './CellLabel/CellLabel';

type Props = {
  arrow?: string;
  clickHandler: (row: number, column: number) => void;
  column: number;
  highlighted?: boolean;
  letter?: string;
  number?: number;
  row: number;
  selected?: boolean;
};
const Cell = ({
  row,
  number,
  column,
  letter,
  arrow,
  clickHandler,
  highlighted,
  selected,
}: Props) => (
  <g onClick={() => clickHandler(row, column)} data-testid="cell">
    <CellRectangle
      column={column}
      row={row}
      selected={selected}
      highlighted={highlighted}
    />
    {number ? <CellLabel column={column} row={row} number={number} /> : null}
    {arrow && arrow === 'across' ? (
      <ArrowAcross column={column} row={row} />
    ) : null}
    {arrow && arrow === 'down' ? <ArrowDown column={column} row={row} /> : null}
    <CellText row={row} column={column} text={letter} />
  </g>
);

export default Cell;
