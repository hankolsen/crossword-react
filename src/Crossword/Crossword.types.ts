/* istanbul ignore file */
export type CellType = {
  [direction in Direction]: string[];
} & {
  column: number;
  number: number;
  row: number;
  highlighted?: boolean;
  selected?: boolean;
  text?: string;
  arrow?: string;
};

export enum Direction {
  across = 'across',
  down = 'down',
}

export type Selection = {};

export type Position = { x: number; y: number };

export type SeparatorType = {
  direction: Direction;
  position: Position;
  separator: string;
  locations: number[];
  id: string;
};

export type CrosswordListItem = {
  name: string;
  id: string;
};

export type CrosswordListResponse = {
  crosswords: CrosswordListItem[];
};
