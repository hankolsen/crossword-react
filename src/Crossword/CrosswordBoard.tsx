import React from 'react';
import CellInput from 'Cells/CellInput/CellInput';
import Cells from 'Cells/Cells';
import Separators from 'Separators/Separators';
import { useCrossword } from './CrosswordContext';

const CrosswordBoard = () => {
  const { boardWidth, boardHeight, separators } = useCrossword();
  return (
    <div className="crossword-board">
      <svg
        className="crossword__grid"
        viewBox={`0 0 ${boardWidth} ${boardHeight}`}
        fill="#222222"
      >
        <rect
          x="0"
          y="0"
          width={boardWidth}
          height={boardHeight}
          className="crossword__grid-background"
        />
        <Cells />
        <Separators separators={separators} />
      </svg>
      <CellInput />
    </div>
  );
};

export default CrosswordBoard;
