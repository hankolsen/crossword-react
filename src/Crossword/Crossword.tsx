import React from 'react';
import LoadingIndicator from '../LoadingIndicator/LoadingIndicator';
import CrosswordBoard from './CrosswordBoard';
import { useCrossword } from './CrosswordContext';

import './Crossword.css';
import ResetButton from './ResetButton';

const Crossword = () => {
  const { isLoading, name } = useCrossword();

  if (isLoading) {
    return (
      <div className="crossword">
        <h3>Loading crossword</h3>
        <LoadingIndicator />
      </div>
    );
  }

  return (
    <div className="crossword">
      <h3>{name}</h3>
      <div className="crossword-container">
        <CrosswordBoard />
      </div>
      <ResetButton />
    </div>
  );
};

export default Crossword;
