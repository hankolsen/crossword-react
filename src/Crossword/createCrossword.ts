import { CELL_HEIGHT, CELL_WIDTH } from '../config';
import { CellType } from './Crossword.types';
import { toggleDirection } from './crosswordHelper';
import fillCell from './fillCell';

type Entry = {
  id: string;
  direction: 'across' | 'down';
  group: string[];
  length: number;
  position: { x: number; y: number };
  separatorLocations?: { [key: string]: number[] };
  turns?: number[];
};

export type CrosswordType = {
  name: string;
  cells: CellType[][];
  separators?: unknown[];
  numberOfColumns: number;
  numberOfRows: number;
  boardWidth: number;
  boardHeight: number;
  inputWidth: number;
  inputHeight: number;
};

type CrosswordResponse = {
  crossword: {
    entries: Entry[];
    name: string;
    size: { width: number; height: number };
    _id: string;
  };
};

const createCrossword = (crosswordId: string): Promise<CrosswordType> =>
  new Promise((resolve, reject) =>
    fetch(`/.netlify/functions/crosswords/${crosswordId}`)
      .then((response) => response.json())
      .then(({ crossword }: CrosswordResponse) => {
        if (!crossword) {
          reject();
        }
        const numberOfColumns = crossword.size.width;
        const numberOfRows = crossword.size.height;
        const separators: unknown[] = [];

        const cells = Array(numberOfRows)
          .fill(null)
          .map(() => Array(numberOfColumns).fill(null));

        const userData = JSON.parse(
          localStorage.getItem(`kryzz-${crosswordId}`) || 'null',
        );

        crossword.entries.sort(
          (a, b) => a.position.y - b.position.y || a.position.x - b.position.x,
        );

        let number = 1;
        const numbers: { [key: string]: number } = {};
        crossword.entries.forEach(
          ({ direction, position, length, separatorLocations, turns }) => {
            let column = position.x;
            let row = position.y;
            let walkingDirection = direction;
            let turnIndex = 0;
            if (numbers[`${column},${row}`]) {
              number = numbers[`${column},${row}`];
            }
            const id = `${number}-${direction}`;
            numbers[`${column},${row}`] = number;
            Array(length)
              .fill(null)
              .forEach((_, i) => {
                const text = userData && userData[row][column];
                fillCell({
                  cells,
                  row,
                  column,
                  index: i,
                  number,
                  id,
                  direction,
                  text,
                  hasTurn: turns,
                });

                if (turns && turns.length && turns[turnIndex] - 1 === i) {
                  walkingDirection = toggleDirection(walkingDirection);
                  cells[row][column].arrow = walkingDirection;
                  turnIndex += 1;
                }

                if (walkingDirection === 'across') {
                  column += 1;
                } else {
                  row += 1;
                }
              });

            Object.entries(separatorLocations || {}).forEach(
              ([separator, locations]) => {
                if (locations && locations.length) {
                  separators.push({
                    direction,
                    position,
                    separator,
                    locations,
                    id,
                  });
                }
              },
            );
            number += 1;
          },
        );

        const inputWidth = 100 / numberOfColumns;
        const inputHeight = 100 / numberOfRows;
        const boardWidth =
          CELL_WIDTH * numberOfColumns + numberOfColumns + 1 || 0;
        const boardHeight = CELL_HEIGHT * numberOfRows + numberOfRows + 1 || 0;
        const { name } = crossword;
        console.log('cells', cells);
        // eslint-disable-next-line max-len
        resolve({
          name,
          cells,
          separators,
          numberOfColumns,
          numberOfRows,
          boardWidth,
          boardHeight,
          inputWidth,
          inputHeight,
        });
      })
      .catch((e) => reject(e)),
  );

export default createCrossword;
