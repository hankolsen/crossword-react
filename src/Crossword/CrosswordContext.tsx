/* eslint-disable @typescript-eslint/no-unused-vars */
import React, {
  createContext,
  FunctionComponent,
  useContext,
  useEffect,
  useState,
} from 'react';
import { useParams } from 'react-router-dom';
import createCrossword, { CrosswordType } from './createCrossword';

type CrosswordContextType = {
  clickHandler: (row: number, column: number) => void;
  crossword?: CrosswordType;
  currentCell: { row: number; column: number };
  isLoading: boolean;
  inputClickHandler: () => void;
  inputHandler: (event: React.KeyboardEvent<HTMLInputElement>) => void;
  keyUpHandler: (event: React.KeyboardEvent<HTMLInputElement>) => void;
  showInput: boolean;
};

const CrosswordContext = createContext<CrosswordContextType | undefined>(
  undefined,
);

const useCrossword = () => {
  const context = useContext(CrosswordContext);
  if (!context) {
    throw new Error('useCrossword must be used within a CrosswordContext');
  }
  return context;
};

const CrosswordProvider: FunctionComponent = ({ children }) => {
  const [crossword, setCrossword] = useState<CrosswordType>();
  const [isLoading, setIsLoading] = useState(true);
  const [showInput, setShowInput] = useState(false);
  const [currentCell, setCurrentCell] = useState({ row: 0, column: 0 });
  const { id } = useParams<{ id: string }>();

  const clickHandler = (row: number, column: number) => {
    if (crossword) {
      setCurrentCell(crossword.cells[row][column]);
    }
    setShowInput(true);
  };

  const inputHandler = (event: React.KeyboardEvent<HTMLInputElement>) => {};
  const keyUpHandler = (event: React.KeyboardEvent<HTMLInputElement>) => {};
  const inputClickHandler = () => {};

  useEffect(() => {
    setIsLoading(true);
    createCrossword(id)
      .then((crossword) => setCrossword(crossword))
      .catch()
      .finally(() => setIsLoading(false));
  }, [id]);

  const value = {
    clickHandler,
    crossword,
    currentCell,
    isLoading,
    inputClickHandler,
    inputHandler,
    keyUpHandler,
    showInput,
  };

  return (
    <CrosswordContext.Provider value={value}>
      {children}
    </CrosswordContext.Provider>
  );
};

export { CrosswordProvider, useCrossword };
