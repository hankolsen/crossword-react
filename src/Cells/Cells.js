import React from 'react';
import { useCrossword } from 'Crossword/CrosswordContext';
import Cell from './Cell';

const Cells = ({ cells }) => {
  const { clickHandler } = useCrossword();

  return cells.map((row, y) =>
    row.map(
      (cell, x) =>
        cell && (
          <Cell
            key={`${y + 1}-${x + 1}`}
            row={y}
            column={x}
            number={cell.number}
            letter={cell.text}
            highlighted={cell.highlighted}
            selected={cell.selected}
            arrow={cell.arrow}
            clickHandler={clickHandler}
          />
        ),
    ),
  );
};

export default Cells;
