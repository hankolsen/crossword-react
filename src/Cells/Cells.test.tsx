import React from 'react';
import { useCrossword } from '../Crossword/CrosswordContext';
import Cells from './Cells';
import { renderWithSvg, screen } from '../test/test-utils';

jest.mock('../Crossword/CrosswordContext');
const mockUseCrossword = useCrossword as jest.Mock;

describe('Cells test', () => {
  it('should not render if there are no cells', () => {
    mockUseCrossword.mockImplementation(() => ({ cells: undefined }));
    const { container } = renderWithSvg(<Cells />);
    expect(screen.queryByTestId('cell')).not.toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });

  it('should not render cells', () => {
    mockUseCrossword.mockImplementation(() => ({
      cells: [[{ number: 2, x: 0, y: 1 }], [{ text: 'q', x: 2, y: 4 }]],
    }));
    const { container } = renderWithSvg(<Cells />);
    expect(screen.queryAllByTestId('cell')).toHaveLength(2);
    expect(screen.queryByText(/^2$/)).toBeInTheDocument();
    expect(screen.queryByText(/^q$/i)).toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });
});
