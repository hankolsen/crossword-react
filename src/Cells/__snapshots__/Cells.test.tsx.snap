// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cells test should not render cells 1`] = `
<div>
  <svg>
    <g
      data-testid="cell"
    >
      <rect
        class="crossword__cell  "
        height="31"
        width="31"
        x="1"
        y="1"
      />
      <text
        class="crossword__cell-number"
        x="2"
        y="10"
      >
        2
      </text>
      <text
        class="crossword__cell-text"
        text-anchor="middle"
        x="15.5"
        y="25"
      />
    </g>
    <g
      data-testid="cell"
    >
      <rect
        class="crossword__cell  "
        height="31"
        width="31"
        x="1"
        y="33"
      />
      <text
        class="crossword__cell-text"
        text-anchor="middle"
        x="15.5"
        y="57"
      >
        q
      </text>
    </g>
  </svg>
</div>
`;

exports[`Cells test should not render if there are no cells 1`] = `
<div>
  <svg />
</div>
`;
